### PR TITLE
Pv2 1868 make period end start wait for submission jobs to complete

### PR DESCRIPTION
**Pull request by keenesolutions**

### CODE REVIEW

File: src/SFA.DAS.Payments.PeriodEnd.Application.UnitTests/PeriodEndJobContextMessageHandlerTests.cs [hunk 1]
```diff
@@ -388,7 +388,7 @@ public async Task Fails_If_No_Collection_Year()
         }
 
         [Test]
-        [TestCase("PeriodEndStart", 0)]
+        [TestCase("PeriodEndStart", 1)]
```
Comment (keenesolutions) [inline src/SFA.DAS.Payments.PeriodEnd.Application.UnitTests/PeriodEndJobContextMessageHandlerTests.cs:391]: Broken test

File: src/SFA.DAS.Payments.Monitoring.Jobs.Application/JobProcessing/JobStatusService.cs [hunk 2]
```diff
@@ -65,6 +67,12 @@ public async Task<bool> ManageStatus(long jobId, CancellationToken cancellationT
                     return true;
             }
 
+            var additionalJobChecksResult = await PerformAdditionalJobChecks(job, cancellationToken);
```
Comment (hardik-desai-kcl) [inline src/SFA.DAS.Payments.Monitoring.Jobs.Application/JobProcessing/JobStatusService.cs:70]: any reason why this is done at this point, could we move it closer to its usage? near line 98

File: src/SFA.DAS.Payments.Monitoring.Jobs.Application.UnitTests/JobProcessing/PeriodEnd/PeriodEndStartJobStatusServiceTests.cs [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    [TestFixture]
+    public class PeriodEndStartJobStatusServiceTests
+    {
+         private AutoMock mocker;
+       
+        private JobModel job;
+        private List<CompletedMessage> completedMessages;
+        private List<InProgressMessage> inProgressMessages;
+
+        private List<(long? DcJobId, JobStatus JobStatus, bool? DcJobSucceeded, DateTimeOffset? endTime)>
```
[the remark this answers is not in the document]
Comment (keenesolutions) [reply src/SFA.DAS.Payments.Monitoring.Jobs.Application.UnitTests/JobProcessing/PeriodEnd/PeriodEndStartJobStatusServiceTests.cs:25]: I've updated this to use a new DTO

File: src/SFA.DAS.Payments.Monitoring.AcceptanceTests/Jobs/PeriodEndJobs.feature [hunk 4]
```diff
@@ -27,4 +30,24 @@ Scenario: Provider Period End Stop  Job Completed
...
+	Given the earnings event service has received and is processing a provider earnings job
+	And the period end service has received a period end start job
+	When the period end service notifies the job monitoring service to record the start job
+	And the final messages for the job are successfully processed for the Period End Start job
+	Then the period end job should not complete
+	And when the final messages for the job are successfully processed for the submission job
+	Then the job monitoring service should update the status of the job to show that it has completed	
+	And the monitoring service should notify other services that the period end start job has completed successfully
+
+#Scenario: Provider Period End Start Job fails if outstanding submissions time out
```
[the remark this answers is not in the document]
Comment (keenesolutions) [reply src/SFA.DAS.Payments.Monitoring.AcceptanceTests/Jobs/PeriodEndJobs.feature:46]: This has been uncommented

File: src/SFA.DAS.Payments.Monitoring.Jobs.Application.UnitTests/JobProcessing/PeriodEnd/PeriodEndStartJobStatusServiceTests.cs [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+                .SetupGet(cfg => cfg.EarningsJobTimeout)
+                .Returns(TimeSpan.FromSeconds(20));
+
+            mocker.Mock<IJobsDataContext>()
+                .Setup(x => x.GetOutstandingOrTimedOutJobs(It.IsAny<long>(), It.IsAny<DateTimeOffset>(), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(outstandingOrTimedOutJobs);
+            
+        }
+
+        //[Test]
```
[the remark this answers is not in the document]
Comment (keenesolutions) [reply src/SFA.DAS.Payments.Monitoring.Jobs.Application.UnitTests/JobProcessing/PeriodEnd/PeriodEndStartJobStatusServiceTests.cs:73]: This has been uncommented

File: src/SFA.DAS.Payments.Monitoring.Jobs.Application.UnitTests/PeriodEndStartJobStatusManagerTests.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+          _ =   statusManager.GetJobStatusService(mockScope.Object);
+          mockScope.Verify(x => x.Resolve<IPeriodEndStartJobStatusService>(), Times.Once);
+        }
+
+        [Test]
+        public async Task GetCurrentJobs_Return_JobsForPeriodEndStart()
+        {
+            var mockStorageService = new Mock<IJobStorageService>();
+            _ = await statusManager.GetCurrentJobs(mockStorageService.Object);
+            mockStorageService.Verify(x=>x.GetCurrentPeriodEndStartJobs(CancellationToken.None), Times.Once);
```
[the remark this answers is not in the document]
Comment (keenesolutions) [reply src/SFA.DAS.Payments.Monitoring.Jobs.Application.UnitTests/PeriodEndStartJobStatusManagerTests.cs:42]: I've added this check.

File: src/SFA.DAS.Payments.Monitoring.Jobs.Application/JobProcessing/JobStatusService.cs (hunk 2 above)
Comment (keenesolutions) [reply src/SFA.DAS.Payments.Monitoring.Jobs.Application/JobProcessing/JobStatusService.cs:70]: Placing it here means that we are sure that all submission jobs are competed before we consider the start jobs own messages.

File: src/SFA.DAS.Payments.Monitoring.Jobs.Application/JobProcessing/JobStatusService.cs [hunk 7]
```diff
@@ -87,11 +95,21 @@ await ManageMessageStatus(jobId, completedItems, inProgressMessages, cancellatio
...
+
+            return await CompleteJob(jobId, jobStatus,
                 currentJobStatus.endTime.Value, cancellationToken).ConfigureAwait(false);
         }
 
+
+        protected virtual Task<(bool IsComplete, JobStatus? OverriddenJobStatus)> PerformAdditionalJobChecks(JobModel job, 
+            CancellationToken cancellationToken)
+        {
+            return Task.FromResult((true,(JobStatus?) null));
```
[the remark this answers is not in the document]
Comment (keenesolutions) [reply src/SFA.DAS.Payments.Monitoring.Jobs.Application/JobProcessing/JobStatusService.cs:108]: It's not returning success but iscomplete = true. We're not going to override the job status so this is returning null.

File: src/SFA.DAS.Payments.Monitoring.Jobs.Application/JobProcessing/PeriodEnd/PeriodEndStartJobStatusService.cs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            this.context = context ?? throw new ArgumentNullException(nameof(context));
+        }
+       
+        protected override async Task<(bool IsComplete, JobStatus? OverriddenJobStatus)> PerformAdditionalJobChecks(JobModel job,  CancellationToken cancellationToken)
+        {
+            var outstandingJobs =
+                await context.GetOutstandingOrTimedOutJobs(job.DcJobId, job.StartTime, cancellationToken);
+            
+            //var timeoutsPresent = outstandingJobs.Any(x =>
```
[the remark this answers is not in the document]
Comment (keenesolutions) [reply src/SFA.DAS.Payments.Monitoring.Jobs.Application/JobProcessing/PeriodEnd/PeriodEndStartJobStatusService.cs:30]: This has been re-enabled.

File: src/SFA.DAS.Payments.Monitoring.Jobs.Data/JobsDataContext.cs [hunk 9]
```diff
@@ -148,5 +149,20 @@ public async Task SaveJobStatus(long dcJobId, JobStatus status, DateTimeOffset e
             job.Status = status;
             await SaveChangesAsync(cancellationToken).ConfigureAwait(false);
         }
+
+        public async Task<List<(long? DcJobId, JobStatus JobStatus, bool? DcJobSucceeded, DateTimeOffset? endTime)>>GetOutstandingOrTimedOutJobs(long? dcJobId,
```
[the remark this answers is not in the document]
Comment (keenesolutions) [reply src/SFA.DAS.Payments.Monitoring.Jobs.Data/JobsDataContext.cs:153]: Done.

File: src/SFA.DAS.Payments.Monitoring.Jobs.Application/JobProcessing/JobStatusService.cs (hunk 2 above)
Comment (keenesolutions) [reply src/SFA.DAS.Payments.Monitoring.Jobs.Application/JobProcessing/JobStatusService.cs:70]: The other way round was causing issues where there were no longer any completed or in progress messages and it ended up just waiting and timing out. This seems like the least invasive way or adding the new addtional checks without impacting the other job status services.

**Pull request merged**